models/interactions.sql

```diff
@@ -0,0 +1,195 @@
+    e.domain_userid,
+    coalesce(y.player_id, me.html_id, r.content_title) as media_id, -- replace content_title with id field when available
+    round(mp.duration) as duration,
+    coalesce(y.player_id, me.html_id, r.content_title) as title, -- replace ids with title fields when available
+    case when me.media_type = 'audio' then 'audio' else 'video' end as media_type,
+    coalesce(y.schema_vendor||'-'||y.schema_name, me.schema_vendor||'-'||me.schema_name, r.schema_vendor||'-'||r.schema_name)  as media_player_type,
+    e.page_referrer,
+    coalesce(e.page_url, r.content_url) as content_url,
+    e.geo_region_name,
+    e.br_name,
+    e.dvce_type,
+    e.os_name,
+    e.os_timezone,
+    mpe.type as event_type,
+    e.derived_tstamp as start_tstamp, 
+    mp.current_time as media_start_pos,
+    isnull(mp.playback_rate, 1) as playback_rate,
+    coalesce(y.playback_quality, ve.video_height||'x'||ve.video_width, rd.display_height||'x'||rd.display_width) as playback_quality,
+    case when event_type = 'ended' then 100 else mp.percent_progress end percent_progress,
+    mp.muted
+ 		
+		from {{ ref("mp_events") }} as e
+
+    left join {{ source('atomic', 'com_snowplowanalytics_snowplow_media_player_event_1') }} as mpe
+    on mpe.root_id = e.event_id
+
+    left join {{ source('atomic', 'com_snowplowanalytics_snowplow_media_player_1') }} as mp
+    on mp.root_id = e.event_id
+
+    left join {{ source('atomic', 'com_youtube_youtube_1') }} as y
+    on  y.root_id = e.event_id
+
+    left join {{ source('atomic', 'org_whatwg_media_element_1') }} as me
+    on me.root_id = e.event_id
+
+    left join {{ source('atomic', 'org_whatwg_video_element_1') }} as ve
+    on ve.root_id = e.event_id
+
+    left join {{ source('atomic', 'com_roku_video_1') }} as r
+    on r.root_id = e.event_id
+
+    left join {{ source('atomic', 'com_roku_device_info_1') }} as rd
+	  on  rd.root_id = e.event_id
+
+)
+
+, interaction_ends as (
+
+	select
+		page_view_id,
+		media_id,
+		event_id,
+		event_type,
+		start_tstamp,
+		media_start_pos,
+		lead(start_tstamp, 1) over(partition by page_view_id, media_id order by start_tstamp) end_tstamp,
+		lead(media_start_pos, 1) over(partition by page_view_id, media_id order by start_tstamp) media_end_pos	
+		
+	from prep	
+
+)
+
+, find_nulls_round_1 as (
+
+	select 
+		event_id,
+    start_tstamp,
+    event_type,
+    case when event_type in ('pause', 'paused', 'ended') then coalesce(end_tstamp, start_tstamp) else end_tstamp end end_tstamp,
+    media_start_pos,
+		case when event_type in ('pause', 'paused', 'ended') then coalesce(media_end_pos, media_start_pos) else media_end_pos end media_end_pos
+	
+	from interaction_ends
+
+)
+
+-- for now default it to start_tstamp, same as above but it could be the page_view end_tstamp
+, find_nulls_round_2 as (
+
+	select
+		f.event_id,
+    f.event_type,
+    f.start_tstamp,
+		case when f.end_tstamp is null then p.start_tstamp else f.end_tstamp end end_tstamp,
```

Comment:
I tried to highlight it in the comment but essentially part2 was just a placeholder to eventually calculate the end timestamps of events that are not a `paused` or `ended` event and take them from the web model`'s pageview table`'s `end_tstamp`s to get a more accurate view. Not sure if it is worth the extra calculation and dependence though.

Reply:
Yeh I guess you could consider either taking the `end_tstamp` from the page views table or use the page ping events to calculate the `end_tstamp` within this model

Reply:
I left it with the simplified version, meaning I will equal the `end_timestamp` of the last event as the `start_tstamp` because as I found that when I added the `end_tstamps` from the `page view` table, the values I got was less than the `start_tstamp` of the last event.